### PR TITLE
feature: password live validation & username compatibility tweaks

### PR DESCRIPTION
**Pull request by alfonsobries**
## Summary

- Adds a live password validation UI to register and update password form
- Fixed some errors related to using usernames instead of emails in login and register form

## Checklist

- [ ] Documentation _(if necessary)_
- [x] Tests _(if necessary)_
- [ ] Ready to be merged

image

### CODE REVIEW

File: src/Components/RegisterForm.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+class RegisterForm extends Component
+{
+    use ValidatesPassword;
+    public string $name = '';
+    public string $username = '';
+    public string $email = '';
+    public string $password = '';
+    public string $password_confirmation = '';
+    public bool $terms = false;
```
Comment (faustbrian) [inline src/Components/RegisterForm.php]: Those should be kept in a `$state` array like all other components do.

https://ark.dev/docs/program-incentives/guidelines/laravel-and-php#state-management

File: tests/MediaUser.php [hunk 2]
```diff
@@ -17,6 +17,7 @@ public static function fake(): self
     {
         return static::create([
             'name'              => 'John Doe',
+            'username'             => 'john.doe',
```
Comment (faustbrian) [inline tests/MediaUser.php]: Formatting

File: tests/Helpers.php [hunk 3]
```diff
@@ -9,6 +9,7 @@ function createUserModel()
 {
     return User::create([
         'name'              => 'John Doe',
+        'username'             => 'john.doe',
```
Comment (faustbrian) [inline tests/Helpers.php]: Formatting

File: resources/lang/en/forms.php [hunk 4]
```diff
@@ -15,4 +15,11 @@
         'requirements_notice' => 'Password must be 12–128 characters, and include a number, a symbol, a lower and an upper case letter.',
     ],
 
+    'password_rules' => [
+        'needs_lowercase' => 'One lowercase character',
+        'needs_uppercase' => 'One uppercase character',
+        'needs_a_number' => 'One number',
+        'needs_special_character' => 'One special characater',
```
Comment (ItsANameToo) [inline resources/lang/en/forms.php]:
```suggestion
        'needs_special_character' => 'One special character',
```

File: tests/Actions/CreateNewUserTest.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => 'sec$%Asfhhdfhfdhgd',
+                'password_confirmation' => 'sec$%Asfhhdfhfdhgd',
+                'terms' => true
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('password'));
+    }
+});
+
+it('password require one special character', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require one special character', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => 'sec$r2t12345',
+                'password_confirmation' => 'sec$r2t12345',
+                'terms' => true
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('password'));
+    }
+});
+
+it('password require one number', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require one number', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => 'Sec$r2t',
+                'password_confirmation' => 'Sec$r2t',
+                'terms' => true
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('password'));
+    }
+});
+
+it('password require an upercase letter', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require an uppercase letter', function () {
```

File: resources/views/auth/login.blade.php [hunk 8]
```diff
@@ -36,7 +36,6 @@ class="flex flex-col p-8 mx-4 border rounded-lg border-theme-secondary-200 md:mx
                     <div class="mb-4">
                         <div class="flex flex-1">
                             <x-ark-input
-                                type="email"
```
Comment (ItsANameToo) [inline resources/views/auth/login.blade.php:39]: why was this removed? 🤔

File: tests/Actions/CreateNewUserTest.php [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => $this->validPassword,
+                'password_confirmation' => 'password',
+                'terms' => false
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('password'));
+    }
+});
+
+it('password must be 12 chars longer', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should be equal to or longer than 12 characters', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => $this->validPassword,
+                'password_confirmation' => $this->validPassword,
+                'terms' => false
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('terms'));
+    }
+});
+
+it('password must match', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should match the confirmation', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => $this->validPassword,
+                'password_confirmation' => $this->validPassword,
+                'terms' => true
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('email'));
+    }
+});
+
+it('must accept terms', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require the terms to be accepted', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => $this->validPassword,
+                'password_confirmation' => $this->validPassword,
+                'terms' => true
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('email'));
+    }
+});
+
+it('requires a valid email', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require a valid email', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+                'password' => $this->validPassword,
+                'password_confirmation' => $this->validPassword,
+                'terms' => true
+            ]);
+    } catch (ValidationException $e) {
+        $this->assertTrue($e->validator->errors()->has('username'));
+    }
+});
+
+it('requires an email', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require an email', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        'terms' => true
+    ]);
+
+    $this->assertEquals('alfonsobries', $user->username);
+    $this->assertEquals('alfonso@ark.io', $user->email);
+    $this->assertEquals('Alfonso Bribiesca', $user->name);
+    $this->assertTrue(Hash::check($this->validPassword, $user->password));
+});
+
+it('requires an username', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should require a username', function () {
```

File: tests/Actions/CreateNewUserTest.php [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+use ARKEcosystem\Fortify\Actions\CreateNewUser;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\Hash;
+use Illuminate\Validation\ValidationException;
+
+beforeEach(function () {
+    $this->validPassword = 'Pas3w05d&123456';
+});
+
+it('creates a valid user with the create user action', function () {
```
Comment (faustbrian) [inline tests/Actions/CreateNewUserTest.php]:
```suggestion
it('should create a valid user with the create user action', function () {
```

**Review by ItsANameToo — DISMISSED** [review]

File: src/FortifyServiceProvider.php [hunk 16]
```diff
@@ -99,6 +103,10 @@ public function registerComponents(): void
         Livewire::component('profile.update-profile-information-form', UpdateProfileInformationForm::class);
         Livewire::component('profile.update-profile-photo-form', UpdateProfilePhotoForm::class);
         Livewire::component('profile.update-timezone-form', UpdateTimezoneForm::class);
+        Livewire::component('auth.password-validator', PasswordValidator::class);
+        Livewire::component('auth.register-form', RegisterForm::class);
+
+        Blade::component('auth.password-rules', PasswordRules::class);
```
Comment (faustbrian) [inline src/FortifyServiceProvider.php]: Split `registerComponents` into `registerLivewireComponents` and `registerBladeComponents` so we can keep them clearly separated.

File: resources/views/auth/login.blade.php (hunk 8 above)
Comment (alfonsobries) [reply resources/views/auth/login.blade.php:39]: @ItsANameToo The login form accepts email or username. Even the label says that. Currently was only accepting an email

File: src/Components/UpdatePasswordForm.php [hunk 17]
```diff
@@ -44,6 +47,11 @@ public function updatePassword(UpdatesUserPasswords $updater)
         $this->emit('saved');
     }
 
+    public function updatedStatePassword($password)
```
Comment (faustbrian) [inline src/Components/UpdatePasswordForm.php]:
```suggestion
    public function updatedStatePassword(string $password): void
```

File: resources/views/auth/login.blade.php (hunk 8 above)
Comment (ItsANameToo) [reply resources/views/auth/login.blade.php:39]: ah nice catch, probably an inconsistency between deployer and marketsquare that caused that 👍

File: src/Components/PasswordRules.php [hunk 18]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace ARKEcosystem\Fortify\Components;
+
+use Illuminate\View\Component;
+
+class PasswordRules extends Component
```
Comment (faustbrian) [inline src/Components/PasswordRules.php]: Does this have to be a class component if we are not doing anything with the `passwordRules` array inside of this class?

File: src/Components/Concerns/ValidatesPassword.php [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+namespace ARKEcosystem\Fortify\Components\Concerns;
+
+use Laravel\Fortify\Rules\Password;
+
+trait ValidatesPassword
+{
+    public array $passwordRules = [
+        'needsLowercase' => false,
+        'needsUppercase' => false,
+        'needsANumber' => false,
```
Comment (faustbrian) [inline src/Components/Concerns/ValidatesPassword.php]:
```suggestion
        'needsNumeric' => false,
```

File: resources/views/auth/register-form.blade.php [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+
+    <div class="mt-4 text-right">
+        <button type="submit" class="w-full button-primary md:w-auto">
+            Create Account
+        </button>
+    </div>
+
+    <div class="text-center">
+        <div class="pt-4 mt-8 border-t border-theme-secondary-200">
+            Already a member? <a href="/login" class="link">Sign in</a>
```
Comment (faustbrian) [inline resources/views/auth/register-form.blade.php:115]:
```suggestion
            Already a member? <a href="{{ route('login') }}" class="link">Sign in</a>
```

File: src/Components/RegisterForm.php [hunk 21]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace ARKEcosystem\Fortify\Components;
+
+use ARKEcosystem\Fortify\Components\Concerns\ValidatesPassword;
+use ARKEcosystem\Fortify\Models;
+use Domain\Collaborator\Models\Invitation;
+use Livewire\Component;
+
+class RegisterForm extends Component
```
Comment (faustbrian) [inline src/Components/RegisterForm.php:9]: Do we really need the whole register process to be a Livewire component or would it be sufficient to have the password part to be Livewire?

File: resources/lang/en/forms.php [hunk 22]
```diff
@@ -15,4 +15,11 @@
         'requirements_notice' => 'Password must be 12–128 characters, and include a number, a symbol, a lower and an upper case letter.',
     ],
 
+    'password_rules' => [
+        'needs_lowercase' => 'One lowercase character',
+        'needs_uppercase' => 'One uppercase character',
+        'needs_a_number' => 'One number',
```
Comment (faustbrian) [inline resources/lang/en/forms.php]:
```suggestion
        'needs_numeric' => 'One number',
```

File: src/Components/RegisterForm.php (hunk 21 above)
Comment (alfonsobries) [reply src/Components/RegisterForm.php:9]: @faustbrian The first version of this feature was a standalone livewire component just for the password. The problem was that the update password form is another livewire component, and the login form was a regular blade view. I dealt with different problems related to validation messages, live updates, etc., between both forms.

After a few tests, this ends being the simplest and reusable solution. Plus, I put the form logic that used to be in the service provider encapsulated inside this component.

**Review by faustbrian — APPROVED** [review]
